ground/src/main/java/com/google/android/ground/persistence/sync/MediaUploadWorker.kt

```diff
@@ -0,0 +1,173 @@
+import com.google.android.ground.model.mutation.SubmissionMutation
+import com.google.android.ground.model.submission.ValueDelta
+import com.google.android.ground.model.submission.isNotNullOrEmpty
+import com.google.android.ground.model.task.Task
+import com.google.android.ground.persistence.local.room.fields.MutationEntitySyncStatus
+import com.google.android.ground.persistence.remote.RemoteStorageManager
+import com.google.android.ground.repository.MutationRepository
+import com.google.android.ground.repository.UserMediaRepository
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import java.io.FileNotFoundException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+/**
+ * A worker that uploads media associated with submissions to the remote storage in the background.
+ *
+ * This worker should only run when the device has a network connection.
+ *
+ * The worker checks the local storage for any submission mutations that are ready for media upload.
+ * If one or more media associated with the submission do not exist locally, the upload fails and is
+ * marked as such. Otherwise, transient failures are considered eligible for retry.
+ */
+@HiltWorker
+class MediaUploadWorker
+@AssistedInject
+constructor(
+  @Assisted context: Context,
+  @Assisted workerParams: WorkerParameters,
+  private val remoteStorageManager: RemoteStorageManager,
+  private val mutationRepository: MutationRepository,
+  private val userMediaRepository: UserMediaRepository,
+) : CoroutineWorker(context, workerParams) {
+
+  private val loiId: String = workerParams.inputData.getString(LOI_ID) ?: ""
+
+  /**
+   * Left-biased combination of two results. Failures propagate. Success will only be returned in
+   * the case that both results are successful.
+   */
+  private fun <T> join(a: kotlin.Result<T>, b: kotlin.Result<T>): kotlin.Result<T> =
+    if (a.isSuccess) b else a
+
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) { doWorkInternal() }
+
+  /** Performs media uploads for all eligible submission mutations associated with a given LOI. */
+  private suspend fun doWorkInternal(): Result {
+    check(loiId.isNotEmpty()) { "work was queued for an empty location of interest ID" }
+    Timber.d("Starting media upload for LOI: $loiId")
+
+    // Set of mutations eligible for media upload.
+    val mutations =
+      mutationRepository
+        .getMutations(loiId, MutationEntitySyncStatus.MEDIA_UPLOAD_PENDING)
+        .filterIsInstance<SubmissionMutation>()
+        .filter { it.retryCount < MAX_RETRY_COUNT }
+        .map {
+          it.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_IN_PROGRESS,
+            retryCount = it.retryCount + 1
+          )
+        }
+    mutationRepository.updateMutations(mutations) // Mark MEDIA_UPLOAD_IN_PROGRESS
+
+    val updatedMutations = mutations.map { uploadSubmissionMediaAndUpdateSubmission(it) }
```

Comment:
Can we refactor logic a bit in this class to make it more self-descriptive? For ex, why do we get `updatedMutations` form the method that uploads and updates submissions, and not from the method called `updateMutations`?

Reply:
Tried to make this a bit clearer by factoring out that line into a separate function. I think some of this can be clarified via improvements to the mutationRepo as well! Thanks!